File: implementations/ios-sdk/swiftui/Screens/MainScreen.swift
import ContentfulOptimization
import SwiftUI

struct MainScreen: View {
    @EnvironmentObject var client: OptimizationClient
    @State private var entries: [[String: Any]] = []
    @State private var showNavigationTest = false
    @State private var showLiveUpdatesTest = false
    @State private var flagSubscribed = false

    /// Derived from the SDK profile so a rehydrated identified profile renders
    /// the reset control after a cold start, and so the control only flips once
    /// `identify` has actually resolved and been persisted.
    private var isIdentified: Bool {
        let traits = client.state.profile?["traits"] as? [String: Any]
        return traits?["identified"] as? Bool == true
    }

    private var networkControlsEnabled: Bool {
        ProcessInfo.processInfo.arguments.contains("--enable-network-controls")
    }

    var body: some View {
        if showNavigationTest {
            NavigationTestScreen(onClose: { showNavigationTest = false })
        } else if showLiveUpdatesTest {
            LiveUpdatesTestScreen(onClose: { showLiveUpdatesTest = false })
        } else {
            mainContent
        }
    }

    private var mainContent: some View {
        VStack {
            HStack {
                if !isIdentified {
                    Button("Identify") { handleIdentify() }
                        .accessibilityIdentifier("identify-button")
                } else {
                    Button("Reset") { handleReset() }
                        .accessibilityIdentifier("reset-button")
                }
                Button("Navigation Test") { showNavigationTest = true }
                    .accessibilityIdentifier("navigation-test-button")
                Button("Live Updates Test") { showLiveUpdatesTest = true }
                    .accessibilityIdentifier("live-updates-test-button")
            }
            .padding()

            // Test-only runtime network controls. XCUITest cannot toggle real
            // connectivity, so the offline-behavior suite drives the SDK online
            // state on the live process — keeping the in-memory Experience queue
            // intact across the offline/online transition.
            if networkControlsEnabled {
                HStack {
                    Button("Go Offline") { client.setOnline(false) }
                        .accessibilityIdentifier("simulate-offline-button")
                    Button("Go Online") { client.setOnline(true) }
                        .accessibilityIdentifier("simulate-online-button")
                }
            }

            if entries.isEmpty {
                Text("Loading...")
            } else {
                OptimizationScrollView(accessibilityIdentifier: "main-scroll-view") {
                    VStack {
                        ForEach(0..<entries.count, id: \.self) { index in
                            let entry = entries[index]
                            if isNestedContent(entry) {
                                NestedContentEntryView(entry: entry)
                            } else {
                                ContentEntryView(entry: entry)
                            }
                        }
                        AnalyticsEventDisplay()
                    }
                }
            }
        }
        .task {
            // Subscribe the event store before any event can fire. The flag-view
            // event emits synchronously on `subscribeToFlag`, so a later
            // subscription (e.g. in a child view's onAppear) would miss it —
            // `eventPublisher` is a PassthroughSubject and does not buffer.
            EventStore.shared.subscribe(to: client.eventPublisher)
            client.consent(true)
            _ = try? await client.page(properties: ["url": "app"])
        }
        .onReceive(
            client.$state
                .map(\.profile)
                .removeDuplicates { lhs, rhs in
                    let options: JSONSerialization.WritingOptions = [.sortedKeys]
                    let lhsData = lhs.flatMap { try? JSONSerialization.data(withJSONObject: $0, options: options) }
                    let rhsData = rhs.flatMap { try? JSONSerialization.data(withJSONObject: $0, options: options) }
                    return lhsData == rhsData
                }
        ) { profile in
            guard profile != nil else { return }
            // Subscribe to the `boolean` flag once a profile (and consent) is
            // available so a flag-view `component` event is emitted — mirrors
            // the React Native app's gated `sdk.states.flag(...).subscribe(...)`.
            if !flagSubscribed {
                flagSubscribed = true
                client.subscribeToFlag("boolean")
            }
            Task {
                entries = await ContentfulFetcher.fetchEntries(ids: AppConfig.entryIds)
            }
        }
    }

    private func isNestedContent(_ entry: [String: Any]) -> Bool {
        guard let sys = entry["sys"] as? [String: Any],
              let contentType = sys["contentType"] as? [String: Any],
              let innerSys = contentType["sys"] as? [String: Any],
              let id = innerSys["id"] as? String
        else { return false }
        return id == "nestedContent"
    }

    private func handleIdentify() {
        Task {
            _ = try? await client.identify(userId: "charles", traits: ["identified": true])
        }
    }

    private func handleReset() {
        client.reset()
        Task {
            _ = try? await client.page(properties: ["url": "app"])
        }
    }
}

File: implementations/android-sdk/compose/src/main/kotlin/com/contentful/optimization/app/screens/MainScreen.kt
package com.contentful.optimization.app.screens

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.contentful.optimization.shared.AppConfig
import com.contentful.optimization.shared.ContentfulFetcher
import com.contentful.optimization.shared.EventStore
import com.contentful.optimization.app.components.AnalyticsEventDisplay
import com.contentful.optimization.app.components.ContentEntryView
import com.contentful.optimization.app.components.NestedContentEntryView
import com.contentful.optimization.app.components.isNestedContent
import com.contentful.optimization.compose.LocalOptimizationClient
import com.contentful.optimization.compose.LocalScrollContext
import com.contentful.optimization.compose.ScrollContext
import kotlinx.coroutines.launch
import org.json.JSONObject

@Composable
fun MainScreen() {
    val client = LocalOptimizationClient.current
    val state by client.state.collectAsState()
    val scope = rememberCoroutineScope()

    var entries by remember { mutableStateOf<List<Map<String, Any>>>(emptyList()) }
    var showNavigationTest by remember { mutableStateOf(false) }
    var showLiveUpdatesTest by remember { mutableStateOf(false) }
    var flagSubscribed by remember { mutableStateOf(false) }
    var viewportHeight by remember { mutableStateOf(0f) }

    LaunchedEffect(Unit) {
        EventStore.subscribe(client.events, scope)
        client.consent(true)
        try { client.page(mapOf("url" to "app")) } catch (_: Exception) {}
    }

    val profileKey = remember(state.profile) {
        state.profile?.let {
            try { JSONObject(it).toString() } catch (_: Exception) { it.hashCode().toString() }
        }
    }

    val isIdentified = remember(state.profile) {
        @Suppress("UNCHECKED_CAST")
        val traits = state.profile?.get("traits") as? Map<String, Any>
        traits?.get("identified") == true
    }

    LaunchedEffect(profileKey) {
        if (state.profile != null) {
            entries = ContentfulFetcher.fetchEntries(
                AppConfig.entryIds,
                AppConfig.defaultContentfulLocale,
            )
            if (!flagSubscribed) {
                flagSubscribed = true
                client.subscribeToFlag("boolean")
            }
        }
    }

    if (showNavigationTest) {
        NavigationTestScreen(onClose = { showNavigationTest = false })
    } else if (showLiveUpdatesTest) {
        LiveUpdatesTestScreen(onClose = { showLiveUpdatesTest = false })
    } else {
        Column {
            Row(modifier = Modifier.padding(8.dp)) {
                if (!isIdentified) {
                    Button(
                        onClick = {
                            scope.launch {
                                try {
                                    client.identify(
                                        userId = "charles",
                                        traits = mapOf("identified" to true),
                                    )
                                } catch (_: Exception) {}
                            }
                        },
                        modifier = Modifier.testTag("identify-button"),
                    ) { Text("Identify") }
                } else {
                    Button(
                        onClick = {
                            client.reset()
                            scope.launch {
                                try { client.page(mapOf("url" to "app")) } catch (_: Exception) {}
                            }
                        },
                        modifier = Modifier.testTag("reset-button"),
                    ) { Text("Reset") }
                }
                Button(
                    onClick = { showNavigationTest = true },
                    modifier = Modifier.testTag("navigation-test-button"),
                ) { Text("Navigation Test") }
                Button(
                    onClick = { showLiveUpdatesTest = true },
                    modifier = Modifier.testTag("live-updates-test-button"),
                ) { Text("Live Updates Test") }
            }

            if (entries.isEmpty()) {
                Text("Loading...")
            } else {
    
                val scrollContext = remember(viewportHeight) {
                    ScrollContext(scrollY = 0f, viewportHeight = viewportHeight)
                }
                CompositionLocalProvider(LocalScrollContext provides scrollContext) {
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .onGloballyPositioned { viewportHeight = it.size.height.toFloat() },
                    ) {
                        Column(
                            modifier = Modifier
                                .testTag("main-scroll-view")
                                .fillMaxSize()
                                .verticalScroll(rememberScrollState()),
                        ) {
                            entries.forEach { entry ->
                                if (isNestedContent(entry)) {
                                    NestedContentEntryView(entry = entry)
                                } else {
                                    ContentEntryView(entry = entry)
                                }
                            }
                            AnalyticsEventDisplay()
                        }
                    }
                }
            }
        }
    }
}
